Secrets 🔐 PR 3 - Integrate secret handling with db-details

jeff303 opened the pull request:
Add helper function to database model namespace to "swap out" secret values with inserted IDs

Calling this from the pre-update fn

Add pre-insert fn for database model to perform the same

Adding test to database model test ns

Part of epic: #18036

jeff303 committed:
Add secret model

Add Liquibase migration for secret table

Implement bare bones model namespace for secret

Add simple test to ensure secret values are stored and retrieved under an encryption key (or not)

jeff303 committed:
Support encryption of binary data

Update encryption code to be able to operate on byte arrays directly (without dealing in base64 String representation), delegating existing methods to point to these new ones

Adding test

jeff303 committed:
Integrate secret handling with db-details

Add helper function to database model namespace to "wwap out" secret values with inserted IDs

Calling this from the pre-update fn

Add pre-insert fn for database model to perform the same

Adding test to database model test ns

TODO: add test for simply not touching one (ensure the existing data remains)

jeff303 commented:
This is PR 3 for implementing secret support. Builds on top of #17649.

jeff303 committed:
Merging master back in to pick up changes

Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Fix support for source

jeff303 committed:
Add :creator_id to secret model to capture the user who created this secret instance/version

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Fix created_at and updated_at timestamp column types

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Add creator-id and created-at sub properties for each secret in db details

Update test to check for these

This should also be sufficient to break the connection details hash (PR #17307)

jeff303 committed:
Add missing remarks to make migrations linter happy

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Fix namespace decl for database_test.clj

pawit-metabase commented on src/metabase/models/secret.clj:
```diff
@@ -25,6 +27,46 @@
+  [id]
+  (db/select-one Secret :id id {:order-by [[:version :desc]]}))
+
+(defn upsert-secret-value!
```
What's the plan for deletion of secrets/secret versions?

Scenario 1 - I added a data source and decided to remove it from Metabase. I expect the password to be deleted from the DB as well.

Scenario 2 - If I did not enable encryption at first but then later turn it on, I expect the old unencrypted value to be deleted from the DB.

jeff303 replied:
Good points.  I will address in a separate PR to keep these fairly bite sized.

pawit-metabase commented on src/metabase/models/database.clj:
```diff
@@ -69,10 +71,60 @@
+          :else ; no details populated; do nothing
+          database)))
+
 ;; TODO - this logic would make more sense in post-update if such a method existed
```
this comment should be moved down as the `(handle-secrets-changes database)` part does not belong in the hypothetical `post-update`

jeff303 replied:
OK, I think I see what you mean.  See update in latest commit.

pawit-metabase replied:
Yep. That's what I meant.

I'm not yet familiar enough with the code base to review the rest of the PRs so I won't be approving here.

jeff303 committed:
PR feedback: move comment

jeff303 committed:
Support updating secret values as part of encryption key rotation

Add test assertions for secret values in the `rotate-encryption-key!-test`

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Switch to looking at a `-path` suffix for a secret property, to indicate that the secret should have :source :file-path

Update database model code and test accordingly

Fixing unrelated indentation issue in email_test.clj

jeff303 committed:
Fix clj-kondo error

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
PR feedback

jeff303 committed:
Merge branch 'master' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

camsaul reviewed (approved):
works for me

jeff303 committed:
Merge branch 'master' into secrets-3-integrate-db-details
